File: frontend/src/types/event.ts
/**
 * Event types based on backend schema
 */

export type EventType = 
  | 'irrigation' 
  | 'fertilization' 
  | 'nutrients' 
  | 'pesticide' 
  | 'pruning' 
  | 'soil_correction' 
  | 'maintenance' 
  | 'harvest'
  | 'other';

export type EventScope = 'farm' | 'plot' | 'subarea' | 'tree_group';

export type EventTag = 'corrective' | 'preventive' | 'experiment' | 'standard';

export interface Event {
  id: string;
  organization_id: string;
  farm_id: string | null;
  plot_id: string | null;
  row_id: string | null;
  type: EventType;
  scope: EventScope;
  scope_id: string | null;
  scope_name: string | null;
  title: string;
  timestamp: string;
  irrigation_data: IrrigationData | null;
  fertilization_data: FertilizationData | null;
  product_data: ProductData | null;
  notes: string | null;
  operator: string | null;
  team: string | null;
  tags: EventTag[];
  created_by: string;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
  // Joined fields
  plot_name?: string;
  farm_name?: string;
  attachments?: EventAttachment[];
}

export interface IrrigationData {
  duration_minutes?: number;
  water_volume_liters?: number;
  method?: string;
  notes?: string;
}

export interface FertilizationData {
  product_name?: string;
  quantity_kg?: number;
  application_method?: string;
  npk_ratio?: string;
  notes?: string;
}

export interface ProductData {
  product_name?: string;
  active_ingredient?: string;
  quantity?: number;
  unit?: string;
  application_method?: string;
  safety_interval_days?: number;
  notes?: string;
}

export interface EventAttachment {
  id: string;
  event_id: string;
  name: string;
  type: 'photo' | 'document' | 'invoice';
  url: string;
  file_size: number | null;
  mime_type: string | null;
  uploaded_by: string | null;
  created_at: string;
}

export interface EventCreate {
  farm_id?: string;
  plot_id?: string;
  row_id?: string;
  type: EventType;
  scope: EventScope;
  scope_id?: string;
  scope_name?: string;
  title: string;
  timestamp: string;
  irrigation_data?: IrrigationData;
  fertilization_data?: FertilizationData;
  product_data?: ProductData;
  notes?: string;
  operator?: string;
  team?: string;
  tags?: EventTag[];
}

export interface EventUpdate {
  type?: EventType;
  scope?: EventScope;
  scope_id?: string;
  scope_name?: string;
  title?: string;
  timestamp?: string;
  irrigation_data?: IrrigationData;
  fertilization_data?: FertilizationData;
  product_data?: ProductData;
  notes?: string;
  operator?: string;
  team?: string;
  tags?: EventTag[];
}

export interface EventFilters {
  farm_id?: string;
  plot_id?: string;
  type?: EventType;
  scope?: EventScope;
  start_date?: string;
  end_date?: string;
}

File: backend/app/schemas/event.py
"""Schemas de eventos."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class IrrigationData(BaseModel):
    """Dados de irrigacao."""

    duration_minutes: int | None = None
    water_volume_liters: float | None = None
    method: str | None = None
    notes: str | None = None


class FertilizationData(BaseModel):
    """Dados de fertilizacao."""

    product_name: str | None = None
    quantity_kg: float | None = None
    application_method: str | None = None
    npk_ratio: str | None = None
    notes: str | None = None


class ProductData(BaseModel):
    """Dados de produto aplicado."""

    product_name: str | None = None
    active_ingredient: str | None = None
    quantity: float | None = None
    unit: str | None = None
    application_method: str | None = None
    safety_interval_days: int | None = None
    notes: str | None = None


class EventBase(BaseModel):
    """Schema base de evento."""

    farm_id: UUID | None = None
    plot_id: UUID | None = None
    row_id: UUID | None = None
    type: str
    scope: str
    scope_id: UUID | None = None
    scope_name: str | None = None
    title: str
    timestamp: datetime
    irrigation_data: IrrigationData | None = None
    fertilization_data: FertilizationData | None = None
    product_data: ProductData | None = None
    notes: str | None = None
    operator: str | None = None
    team: str | None = None
    tags: list[str] = []


class EventCreate(EventBase):
    """Schema para criacao de evento."""

    pass


class EventUpdate(BaseModel):
    """Schema para atualizacao de evento."""

    type: str | None = None
    scope: str | None = None
    scope_id: UUID | None = None
    scope_name: str | None = None
    title: str | None = None
    timestamp: datetime | None = None
    irrigation_data: IrrigationData | None = None
    fertilization_data: FertilizationData | None = None
    product_data: ProductData | None = None
    notes: str | None = None
    operator: str | None = None
    team: str | None = None
    tags: list[str] | None = None


class EventResponse(EventBase):
    """Schema de resposta de evento."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    created_by: UUID | None = None
    updated_by: UUID | None = None
    created_at: datetime
    updated_at: datetime


class EventAttachmentResponse(BaseModel):
    """Schema de resposta de anexo de evento."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_id: UUID
    name: str
    type: str
    url: str
    file_size: int | None = None
    mime_type: str | None = None
    uploaded_by: UUID | None = None
    created_at: datetime
